### DISCUSSION TITLE
Support `.contains_properly` by expanding `pairwise_point_in_polygon` to support multipolygon input?

### DISCUSSION THREAD

**Discussion by isVoid** [rapidsai/cuspatial#997] [Ideas]
#910 adopts `quadtree_point_in_polygon` to support python's `.contains_properly` API. However, `quadtree_PiP` is in nature all-pairs. There is a complex translation step involved to convert this result to match geopandas output. On the other hand, libcuspatial do support pairwise operation in C++, the missing link is that this API only supports "single polygon array" input. Perhaps we can explore expanding this API to accept multipolygon input and simplify `.contains_properly` implementation?

**Comment by harrism** [comment]
I also would like to explore combining both C++ all-pairs PIP algorithms into a single function that automatically chooses the optimal path (perhaps with an optional quadtree argument to pass in a pre-built index for reuse) and always returns the results in a format that is optimal for the Python UI.

**Comment by harrism** [comment]
I think this should be an issue. I've already decided that this is important to do. :)
